src/operations.jl

```diff
@@ -1,91 +1,107 @@
 import Base.+, Base.-, Base.*, LinearAlgebra.mul!
 
+function mul!(res::AbstractVector, op::AbstractLinearOperator{T}, v::AbstractVector, α, β) where T 
+  (size(v, 1) == size(op, 2) && size(res, 1) == size(op, 1)) || throw(LinearOperatorException("shape mismatch"))
+  increase_nprod(op)
+  op.prod!(res, v, α, β)
+end
+
+function mul!(res::AbstractVector, op::AbstractLinearOperator, v::AbstractVector{T}) where T
+  mul!(res, op, v, one(T), zero(T))
+end
+
 # Apply an operator to a vector.
 function *(op::AbstractLinearOperator{T}, v::AbstractVector{S}) where {T, S}
-  size(v, 1) == size(op, 2) || throw(LinearOperatorException("shape mismatch"))
-  increase_nprod(op)
-  op.prod(v)::typeof(v).name.wrapper{promote_type(T, S), typeof(v).parameters[2]}
+  nrow, ncol = size(op)
+  res = similar(v, promote_type(T, S), nrow)
+  mul!(res, op, v)
+  return res
 end
 
 # Unary operations.
 +(op::AbstractLinearOperator) = op
 
-function -(op::AbstractLinearOperator{T}) where {T}
-  prod = @closure v -> -op.prod(v)
-  tprod = @closure u -> -op.tprod(u)
-  ctprod = @closure w -> -op.ctprod(w)
-  LinearOperator{T}(op.nrow, op.ncol, op.symmetric, op.hermitian, prod, tprod, ctprod)
+function -(op::AbstractLinearOperator{T}) where T
+  prod! = @closure (res, v, α, β) -> mul!(res, op, v, -α, β)
+  tprod! = @closure (res, u, α, β) -> mul!(res, transpose(op), u, -α, β)
+  ctprod! = @closure (res, w, α, β) -> mul!(res, adjoint(op), w, -α, β)
+  LinearOperator{T}(op.nrow, op.ncol, op.symmetric, op.hermitian, prod!, tprod!, ctprod!)
 end
 
-function mul!(y::AbstractVector, op::AbstractLinearOperator, x::AbstractVector)
-  y .= op * x
-  return y
+function prod_op!(res::AbstractVector, op1::AbstractLinearOperator, op2::AbstractLinearOperator, 
+                  vtmp::AbstractVector, v::AbstractVector, α, β)
+  mul!(vtmp, op2, v)
+  mul!(res, op1, vtmp, α, β)
 end
 
-# Binary operations.
+function ctprod_op!(res::AbstractVector, op1::AbstractLinearOperator, op2::AbstractLinearOperator, 
+                    wtmp::AbstractVector, w::AbstractVector, α, β)
+  mul!(wtmp, op1, w)
+  mul!(res, op2, wtmp, α, β)
+end
 
 ## Operator times operator.
 function *(op1::AbstractLinearOperator, op2::AbstractLinearOperator)
+  T = promote_type(eltype(op1), eltype(op2))
   (m1, n1) = size(op1)
   (m2, n2) = size(op2)
   if m2 != n1
     throw(LinearOperatorException("shape mismatch"))
   end
-  S = promote_type(eltype(op1), eltype(op2))
-  prod = @closure v -> op1 * (op2 * v)
-  tprod = @closure u -> transpose(op2) * (transpose(op1) * u)
-  ctprod = @closure w -> op2' * (op1' * w)
-  LinearOperator{S}(m1, n2, false, false, prod, tprod, ctprod)
+  #tmp vector for products
+  vtmp = zeros(T, m2)
+  utmp = zeros(T, n1)
+  wtmp = zeros(T, n1)
+  prod! = @closure (res, v, α, β) -> prod_op!(res, op1, op2, vtmp, v, α, β)
+  tprod! = @closure (res, u, α, β) -> ctprod_op!(res, transpose(op1), transpose(op2), utmp, u, α, β)
```

Comment:
Shoudn’t they be reversed? (AB)’v = B’(A’v).

Reply:
`That's already the case in the function ctprod_op!` no?